### PR TITLE
Generalize repo analysis

### PR DESCRIPTION
**Pull request by DaveCTurner**
In #93825 we introduced a CAS operation on snapshot repositories. In due course we'll want to verify that this operation works as expected using the repository analysis API. This commit reworks the implementation slightly so as to simplify the change which will add this functionality.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.concurrent.Semaphore;
+import java.util.function.BiConsumer;
+
+public class ThrottledIterator<T> implements Releasable {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java:24]: This class (and its test suite) extracted from #93735.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        int maxConcurrency,
+        Runnable onItemCompletion,
+        Runnable onCompletion
+    ) {
+        try (var throttledIterator = new ThrottledIterator<>(iterator, itemConsumer, maxConcurrency, onItemCompletion, onCompletion)) {
+            throttledIterator.run();
+        }
+    }
+
+    private final RefCounted throttleRefs;
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java:56]: nit: I find `throttleRefs` a bit confusing in this context, for what I understand this reference count does not control the throttling, it's used to notify the listener once all the tasks have finished, maybe we can find a clearer name?

**Review by fcofdez — APPROVED** [review]
LGTM, I left a minor comment about a field name. 👍

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java]: Sure, on reflection `refs` seems better to me.

**Pull request merged**